qiskit-superstaq/qiskit_superstaq/superstaq_job.py

```diff
@@ -97,45 +98,91 @@ def result(self, timeout: Optional[float] = None, wait: float = 5) -> qiskit.res
 
-        status = "Done"
+        self._update_status_queue_info()
 
-        # when we have multiple jobs, we will take the "worst status" among the jobs
-        # For example, if any of the jobs are still queued, we report Queued as the status
-        # for the entire batch.
-        for job_id in job_id_list:
-            result = self._backend._provider._client.get_job(job_id)
-            temp_status = result["status"]
+    def _update_status_queue_info(self) -> None:
+        """Updates the overall status based on status queue info.
 
-            if temp_status == "Queued":
-                status = "Queued"
+        Note:
+            When we have multiple jobs, we will take the "worst status" among the jobs.
+            The worst status check follows the chain: Submitted -> Queued -> Running -> Error
+            -> Cancelled -> Done. For example, if any of the jobs are still queued (even if
+            some are done), we report 'Queued' as the overall status of the entire batch.
+        """
+
+        job_id_list = self._job_id.split(",")  # separate aggregated job ids
+
+        status_occurance = Counter(self._job_info[job_id]["status"] for job_id in job_id_list)
+        status_priority_order = ("Submitted", "Queued", "Running", "Error", "Canceled")
+
+        for temp_status in status_priority_order:
+            if status_occurance[temp_status] > 0:
+                self._overall_status = temp_status
                 break
-            elif temp_status == "Running":
-                status = "Running"
 
-        assert status in ["Queued", "Running", "Done"]
+        if status_occurance["Done"] == len(job_id_list):
+            self._overall_status = "Done"
+
+    def status(self) -> qiskit.providers.jobstatus.JobStatus:
+        """Query for the equivalent qiskit job status.
+
+        Returns:
+            The equivalent `qiskit.providers.jobstatus.JobStatus` type.
+        """
 
-        if status == "Queued":
-            status = qiskit.providers.jobstatus.JobStatus.QUEUED
-        elif status == "Running":
-            status = qiskit.providers.jobstatus.JobStatus.RUNNING
-        else:
-            status = qiskit.providers.jobstatus.JobStatus.DONE
-        return status
+        status_match = {
+            "Queued": qiskit.providers.jobstatus.JobStatus.QUEUED,
+            "Running": qiskit.providers.jobstatus.JobStatus.RUNNING,
+            "Submitted": qiskit.providers.jobstatus.JobStatus.INITIALIZING,
+            "Canceled": qiskit.providers.jobstatus.JobStatus.CANCELLED,
+            "Error": qiskit.providers.jobstatus.JobStatus.ERROR,
+            "Done": qiskit.providers.jobstatus.JobStatus.DONE,
+        }
+
+        if self._overall_status in self.TERMINAL_STATES:
+            return status_match.get(self._overall_status)
+
+        self._refresh_job()
+        status = self._overall_status
+
+        assert status in ("Queued", "Running", "Submitted", "Canceled", "Error", "Done")
```

Comment:
is there any chance of this failing? if so we should add more specific handling for that case (e.g. either a more specific exception, or maybe something like an "Unknown" state?). Otherwise we should just remove it.

(in general we should try to avoid assertions client-side)

Reply:
Noted. I don't expect it to fail so I've removed the assertion.